ecp.c no longer builds with Microsoft Visual Studio

### Description
- Type: Bug
- Priority: Blocker

---------------------------------------------------------------
## Bug

**OS**  
Windows

**mbed TLS build:**  
Version: 169712e15ac142412063d3fcae1894ac9e7e9e6c  
Configuration: default
Compiler and options (if you used a pre-built binary, please indicate how you obtained it):  

Microsoft Visual Studio versions 2013, 2015, and 2017 when building Release or Debug builds with CMake

When building the following warning are generated, which are treated as errors and fail the build:

```
         ecp.c
    52>C:\builds\ws\mbedtls-release\worktrees\tmp19i7oxoy\library\ecp.c(1583): error C2220: warning treated as error - no 'object' file generated [C:\builds\ws\mbedtls-release\worktrees\tmp19i7oxoy\cmake_solution\library\mbedcrypto.vcxproj]
    52>C:\builds\ws\mbedtls-release\worktrees\tmp19i7oxoy\library\ecp.c(1583): warning C4267: 'function' : conversion from 'size_t' to 'unsigned int', possible loss of data [C:\builds\ws\mbedtls-release\worktrees\tmp19i7oxoy\cmake_solution\library\mbedcrypto.vcxproj]
    52>C:\builds\ws\mbedtls-release\worktrees\tmp19i7oxoy\library\ecp.c(1620): warning C4267: 'function' : conversion from 'size_t' to 'unsigned int', possible loss of data [C:\builds\ws\mbedtls-release\worktrees\tmp19i7oxoy\cmake_solution\library\mbedcrypto.vcxproj]
```

This was introduced by non-blocking ECC support.

--- a/include/mbedtls/ecp.h
+++ b/include/mbedtls/ecp.h
@@ -233,7 +233,9 @@ int mbedtls_ecp_check_budget( const mbedtls_ecp_group *grp,
                               unsigned ops );
 
 /* Utility macro for checking and updating ops budget */
-#define MBEDTLS_ECP_BUDGET( ops )   MBEDTLS_MPI_CHK( mbedtls_ecp_check_budget( grp, rs_ctx, ops ) );
+#define MBEDTLS_ECP_BUDGET( ops )   \
+    MBEDTLS_MPI_CHK( mbedtls_ecp_check_budget( grp, rs_ctx, \
+                                               (unsigned) (ops) ) );
 
 #else /* MBEDTLS_ECP_RESTARTABLE */
 

--- a/library/ecp.c
+++ b/library/ecp.c
@@ -248,9 +248,16 @@ int mbedtls_ecp_check_budget( const mbedtls_ecp_group *grp,
         else if( grp->pbits >= 384 )
             ops *= 2;
 
-        /* avoid infinite loops: always allow first step */
-        if( rs_ctx->ops_done != 0 && rs_ctx->ops_done + ops > ecp_max_ops )
+        /* Avoid infinite loops: always allow first step.
+         * Because of that, however, it's not generally true
+         * that ops_done <= ecp_max_ops, so the check
+         * ops_done > ecp_max_ops below is mandatory. */
+        if( ( rs_ctx->ops_done != 0 ) &&
+            ( rs_ctx->ops_done > ecp_max_ops ||
+              ops > ecp_max_ops - rs_ctx->ops_done ) )
+        {
             return( MBEDTLS_ERR_ECP_IN_PROGRESS );
+        }
 
         /* update running count */
         rs_ctx->ops_done += ops;

